src/irr/asset/CImageWriterOpenEXR.cpp

```diff
@@ -0,0 +1,159 @@
+#include "openexr/OpenEXR/IlmImf/ImfArray.h"
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "openexr/OpenEXR/IlmImf/ImfNamespace.h"
+namespace IMF = OPENEXR_IMF_NAMESPACE;
+namespace IMATH = IMATH_NAMESPACE;
+
+namespace irr
+{
+	namespace asset
+	{
+		using namespace IMF;
+		using namespace IMATH;
+
+		constexpr uint8_t availableChannels = 4;
+
+		template<typename ilmType>
+		void createAndWriteImage(std::array<ilmType*, availableChannels>& pixelsArrayIlm, const asset::ICPUImage* image, const char* fileName)
+		{
+			auto getIlmType = [&]()
+			{
+				if (image->getCreationParameters().format == EF_R16G16B16A16_SFLOAT)
+					return PixelType::HALF;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_SFLOAT)
+					return PixelType::FLOAT;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_UINT)
+					return PixelType::UINT;
+			};
+
+			Header header(image->getCreationParameters().extent.width, image->getCreationParameters().extent.height);
+			const PixelType pixelType = getIlmType();
+			FrameBuffer frameBuffer;
+
+			const auto& width = image->getCreationParameters().extent.width;
+			const auto& height = image->getCreationParameters().extent.height;
+			const auto pitch = image->getRegions().begin()->bufferRowLength;
+
+			for (uint8_t channel = 0; channel < availableChannels; ++channel)
+				pixelsArrayIlm[channel] = _IRR_NEW_ARRAY(ilmType, width * height);
+
+			for (uint64_t yPos = 0; yPos < height; ++yPos)
+				for (uint64_t xPos = 0; xPos < width; ++xPos)
+				{
+					const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * pitch * availableChannels) + (xPos * availableChannels);
+					const uint64_t ptrStyleIlmShiftToDataChannelPixel = (yPos * width) + xPos;
+
+					for (uint8_t channelIndex = 0; channelIndex < availableChannels; ++channelIndex)
+					{
+						ilmType channelPixel = *(reinterpret_cast<const ilmType*>(image->getBuffer()->getPointer()) + ptrStyleEndShiftToImageDataPixel + channelIndex);
+						*(pixelsArrayIlm[channelIndex] + ptrStyleIlmShiftToDataChannelPixel) = channelPixel;
+					}
+				}
+
+			constexpr std::array<char*, availableChannels> rgbaSignatureAsText = { "R", "G", "B", "A" };
+			for (uint8_t channel = 0; channel < rgbaSignatureAsText.size(); ++channel)
+			{
+				header.channels().insert(rgbaSignatureAsText[channel], Channel(pixelType));
+				frameBuffer.insert
+				(
+					rgbaSignatureAsText[channel],                                                                // name
+					Slice(pixelType,                                                                             // type
+					(char*) pixelsArrayIlm[channel],                                                             // base
+					sizeof(*pixelsArrayIlm[channel]) * 1,                                                        // xStride
+					sizeof(*pixelsArrayIlm[channel]) * image->getCreationParameters().extent.width)              // yStride
+				);
+			}
```

Comment:
same bugs as before.

I'd actually make a Width X Height "framebuffer" (just an array) and copy regions to it manually as necessary.

Reply:
it's done well

>  a Width X Height "framebuffer"

that's what it is
pixelsArrayIlm keeps colors matched to single channel
https://github.com/buildaworldnet/IrrlichtBAW/pull/403#discussion_r361109229

Reply:
important is that I have to insert each channel separately because RGBA class forces you to use half format
that's why I pass to it such an array 

> a Width X Height "framebuffer"

so it's correct

Reply:
if you set up the offset (pointer), stride (penultimate parameter) and row length (last parameter) properly, you'll be able to basically `memcpy` entire rows.

Reply:
so actualy I only have to copy to pixelsArrayIlm pixels correctly having regions taken into account